### PR TITLE
Shapely functions as sample process

### PR DESCRIPTION
**Pull request by jeafreezy**
# Overview

This PR exposes some selected shapely functions as sample process. The selection cut across different operations in shapely. It uses the name of the category of that function as the namespace. E.g union operation under the set module is described as `set:union`. This is to avoid clashes with other functions.

As mentioned in the issue, it accepts list of geometry inputs (WKT and/or GeoJSON geometry), performs the operation and returns the result in the specified **output_format** (optional).

The approach is describe in the image below:

image

# Related issue / discussion

https://github.com/geopython/pygeoapi/issues/1439

# Additional information

**Todo/Idea/Challenge**

- Tests(units) - I didn't see any test for process in the codebase. I tried writing one, but I need to spend some more time to understand the test structure.

- I am yet to figure out how to pass optional function parameters in the inputs. I am still confused on the data type to use in the metadata, especially since different functions require different parameters. e.g **buffer** operation requires **distance**. @tomkralidis and I discussed an alternative approach earlier today, which is using different classes for different functions. That could potentially solve this issue. But I didn't have much time to explore that. Although this may also lead to repetitive code. Since the idea is to have it as a sample, hopefully this current implementation is sufficient.

- Idea: Perhaps in the future, it will be great to chain functions. E.g passing multiple **operations** like ['buffer', 'join'] can be interpreted as perfoming the **buffer** operation on the geometries, followed by the **join** operation before returning the result.

# Dependency policy (RFC2)

- [x] I have ensured that this PR meets [RFC2](https://pygeoapi.io/development/rfc/2) requirements

# Updates to public demo

- [ ] I have ensured that breaking changes to the [pygeoapi master demo server](https://github.com/geopython/demo.pygeoapi.io) have been addressed
  - [ ] https://github.com/geopython/demo.pygeoapi.io/blob/master/services/pygeoapi_master/local.config.yml

# Contributions and licensing

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [x] I'd like to contribute [feature X|bugfix Y|docs|something else] to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution
- [x] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines

### CODE REVIEW

**Comment by francbartoli** [comment]
@jeafreezy thanks for your great contributions. I have just one general comment about security:
- Exposing these functions publicly would leave a big surface for DoS attacks so we should make sure that they are not enabled as default processes sample
- My suggestion is to pick up them from the configuration if additionally configured

**Comment by tomkralidis** [comment]
Agree; it makes sense to have the plugin in the codebase without having in default configuration.

**Comment by jeafreezy** [comment]
Thank you both for the review and feedback. Please, how do you think I should go about this? I have a few ideas;

- Should there be a new entry in the **example-config.yml** for functions, where the user can list the functions they want to enable? Something like below;

image

- Or the shapely-functions entry should be commented out in the config, so they user can enable it, if they want to ? Something like below;

image

- Or is there a better approach to implement the suggestion ?

Kindly advise. Thank you.

**Comment by webb-ben** [comment]
Hi @jeafreezy, I believe the correct approach would be to omit the configuration entirely from the default pygeoapi configuration file, and instead, documenting the capacities of the Shapely Process Provider [here](https://docs.pygeoapi.io/en/latest/data-publishing/ogcapi-processes.html) / [location in source code](https://github.com/geopython/pygeoapi/blob/master/docs/source/data-publishing/ogcapi-processes.rst)

**Comment by jeafreezy** [comment]
Thank you @webb-ben . I will work on that

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -4,16 +4,15 @@
 files: '\.(py|sh|sql)$'
```
Comment (webb-ben) [inline .pre-commit-config.yaml:4]: Are changes to this file necessary?

File: docs/source/data-publishing/ogcapi-processes.rst [hunk 2]
```diff
@@ -9,17 +9,24 @@ fashion (inputs, outputs).
...
+    
+    # enabled by default
        hello-world:
            processor:
                name: HelloWorld
+        
+        #Add this block to the configuration file to enable shapely processes.
+        #shapely-functions:
+           #processor:
+               #name: ShapelyFunctions
```
Comment (webb-ben) [inline docs/source/data-publishing/ogcapi-processes.rst:29]: Could you move this to the start of the Shapely section, with the process uncommented?
```md
Shapely Functions (Optional)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: yaml

        shapely-functions:
           processor:
               name: ShapelyFunctions
```

File: pygeoapi-config.yml [hunk 3]
```diff
@@ -302,3 +302,4 @@ resources:
         type: process
```
Comment (webb-ben) [inline pygeoapi-config.yml:302]: Are changes to this file necessary?

File: pygeoapi/templates/tilematrixsets/tilematrixset.html [hunk 4]
```diff
@@ -1,18 +1,19 @@
-{% extends "_base.html" %}
-{% block title %}{{ super() }} {% trans %}TileMatrixSet{% endtrans %} {% endblock %}
-{% block crumbs %}{{ super() }}
-/ <a href="../TileMatrixSets/">{% trans %}TileMatrixSets{% endtrans %}</a>
-/ <a href="../TileMatrixSets/{{ data['id'] }}?f=html">{{ data['id'] }}</a>
-{% endblock %}
-{% block body %}
+{% extends "_base.html" %} {% block title %}{{ super() }} {% trans
```
Comment (webb-ben) [inline pygeoapi/templates/tilematrixsets/tilematrixset.html:1]: Are changes to this file necessary?

**Review by webb-ben — CHANGES_REQUESTED** [review]
Not sure if we need need pytest to ensure function of the process capacities. 

Apart from that, small changes to docs and a couple of files I am not sure need to be included in this PR, got my +1.

**Comment by jeafreezy** [comment]
> Not sure if we need need pytest to ensure function of the process capacities.
> 
> Apart from that, small changes to docs and a couple of files I am not sure need to be included in this PR, got my +1.

Thank you for reviewing my PR. I will address the issues.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (jeafreezy) [reply .pre-commit-config.yaml:4]: image

I believe this happened when I synced with the main branch recently. I was not my edit.

File: docs/source/data-publishing/ogcapi-processes.rst (hunk 2 above)
Comment (jeafreezy) [reply docs/source/data-publishing/ogcapi-processes.rst:29]: Alright. Will do.

File: pygeoapi-config.yml (hunk 3 above)
Comment (jeafreezy) [reply pygeoapi-config.yml:302]: It's as a result of syncing with the master branch.

File: pygeoapi/templates/tilematrixsets/tilematrixset.html (hunk 4 above)
Comment (jeafreezy) [reply pygeoapi/templates/tilematrixsets/tilematrixset.html]: I didn't changed the files. But I think it was just reformatted by my editor.

File: .gitignore
Comment (tomkralidis) [inline .gitignore:1]: Can we remove this addition?

File: .pre-commit-config.yaml
Comment (tomkralidis) [inline .pre-commit-config.yaml:1]: Can we remove this addition?

File: docs/source/data-publishing/ogcapi-processes.rst [hunk 5]
```diff
@@ -9,14 +9,16 @@ fashion (inputs, outputs).
 pygeoapi implements OGC API - Processes functionality by providing a plugin architecture, thereby
 allowing developers to implement custom processing workflows in Python.
 
-A `sample`_ ``hello-world`` process is provided with the pygeoapi default configuration.
+The pygeoapi offers two processes: a default ``hello-world`` process which allows you to quickly explore the capabilities of processes, and an optional ``shapely-functions`` process with more advanced features that leverages Shapely_ to expose various geometric processing functionality.
```
Comment (tomkralidis) [inline docs/source/data-publishing/ogcapi-processes.rst:12]: `Shapely_` -> `` `Shapely`_``

File: docs/source/data-publishing/ogcapi-processes.rst [hunk 6]
```diff
@@ -107,14 +112,87 @@ Processing examples
...
    # execute a job for the ``hello-world`` process with a success subscriber
+    curl -X POST http://localhost:5000/processes/hello-world/execution \
+        -H "Content-Type: application/json" \
+        -d "{\"inputs\":{\"name\": \"hi there2\"}, \
+            \"subscriber\": {\"successUri\": \"https://www.example.com/success\"}}"
+
+Shapely Functions (Optional)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The `shapely-functions` process exposes some selected Shapely_ functions as sample process. The selection cut across different operations in shapely. To avoid function collision, it uses the name of the function category as the namespace. E.g *union* operation under the *set* module is described as *set:union*.
```
Comment (tomkralidis) [inline docs/source/data-publishing/ogcapi-processes.rst:124]: - `Shapely_` -> `` `Shapely`_``
- shapely -> Shapely

File: pygeoapi-config.yml (hunk 3 above)
Comment (tomkralidis) [reply pygeoapi-config.yml:302]: Can we remove this addition?

File: pygeoapi/process/shapely_functions.py [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# =================================================================
+
+import enum
```
Comment (tomkralidis) [inline pygeoapi/process/shapely_functions.py:32]: Imports should be spaced and ordered as follows:

```python
import enum
import logging

from typing import Any, Dict, List, Tuple, Union

from shapely.geometry import mapping, shape
from shapely.geometry.base import BaseGeometry
from shapely.prepared import prep
from shapely.wkt import loads

from pygeoapi.process.base import BaseProcessor, ProcessorExecuteError
```

File: pygeoapi/process/shapely_functions.py [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+        "predicates",
+        "set operations",
+        "constructive ops",
+    ],
+    "links": [
+        {
+            "type": "text/html",
+            "rel": "about",
+            "title": "information",
+            "href": "https://example.org/process",
```
Comment (tomkralidis) [inline pygeoapi/process/shapely_functions.py:125]: Update link to Shapely, i.e. `https://shapely.readthedocs.io`

File: pygeoapi/templates/tilematrixsets/tilematrixset.html [hunk 9]
```diff
@@ -15,4 +15,4 @@ <h1>{{ data['id'] }} tile matrix set</h1>
   </ul>
 
 </section>
-{% endblock %}
+{% endblock %}
```
Comment (tomkralidis) [inline pygeoapi/templates/tilematrixsets/tilematrixset.html:18]: Can we remove this addition?

File: pygeoapi/process/shapely_functions.py
Comment (tomkralidis) [inline pygeoapi/process/shapely_functions.py:1]: Overall comment: any Python objects with trailing commas (dicts, lists, etc.) should have the trailing commas removed; example:

from

```python
dict_ = {
    'foo': 'bar',
}
```

to


```python
dict_ = {
    'foo': 'bar'
}
```

**Review by tomkralidis — CHANGES_REQUESTED** [review]

**Comment by jeafreezy** [comment]
Thank you for reviewing my PR @tomkralidis. I have addressed the issues to the best of my knowledge. Kindly let me know if there is anything else I missed. Thank you.

File: pygeoapi/process/shapely_functions.py
Comment (jeafreezy) [reply pygeoapi/process/shapely_functions.py:1]: Thank you for the feedback @tomkralidis

**Review by tomkralidis — APPROVED** [review]

**Pull request merged**